File: examples/background_tasks.py
"""Tokio background tasks — usage example.

Process-wide fire-and-forget / delayed jobs (not a durable queue).

    python examples/background_tasks.py

API:

    from fusion_framework import tasks

    tid = tasks.spawn(fn)                 # run ASAP on Tokio
    tid = tasks.spawn_after(ms, fn)       # delay then run
    tasks.cancel(tid)                     # abort pending/running
    tasks.status(tid)                     # pending|running|done|cancelled|failed
"""

from __future__ import annotations

import time

from fusion_framework import tasks


def main() -> None:
    tasks.reset()
    done = {"n": 0}

    def work() -> None:
        done["n"] += 1
        print("  work() ran")

    # 1) Fire-and-forget
    tid = tasks.spawn(work)
    print(f"spawn     id={tid}  status={tasks.status(tid)}")
    for _ in range(50):
        if done["n"] >= 1 and tasks.status(tid) == "done":
            break
        time.sleep(0.02)
    print(f"          status={tasks.status(tid)}  count={done['n']}")

    # 2) Delayed start, then let it finish
    delayed = tasks.spawn_after(150, work)
    print(f"after     id={delayed}  status={tasks.status(delayed)}")
    for _ in range(50):
        if done["n"] >= 2 and tasks.status(delayed) == "done":
            break
        time.sleep(0.02)
    print(f"          status={tasks.status(delayed)}  count={done['n']}")

    # 3) Cancel before the delay elapses
    doomed = tasks.spawn_after(500, work)
    print(f"cancel    id={doomed}  cancelled={tasks.cancel(doomed)}")
    time.sleep(0.2)
    print(f"          status={tasks.status(doomed)}  count={done['n']} (unchanged)")


if __name__ == "__main__":
    main()

File: examples/background_tasks.cs
// Tokio background tasks — usage example.
//
// Process-wide fire-and-forget / delayed jobs (not a durable queue).
//
// Build fusion-ffi, then from repo root (or set FUSION_FFI_PATH):
//   cargo build -p fusion-ffi --release
//   # run via a small console project that references FusionFramework
//
// API:
//   BackgroundTasks.Spawn(Action)
//   BackgroundTasks.SpawnAfter(ms, Action)
//   BackgroundTasks.Cancel(id)
//   BackgroundTasks.Status(id)  // pending|running|done|cancelled|failed

using FusionFramework;

BackgroundTasks.Reset();
var count = 0;
void Work()
{
    Interlocked.Increment(ref count);
    Console.WriteLine("  work() ran");
}

// 1) Fire-and-forget
var tid = BackgroundTasks.Spawn(Work);
Console.WriteLine($"spawn     id={tid}  status={BackgroundTasks.Status(tid)}");
for (var i = 0; i < 50; i++)
{
    if (count >= 1 && BackgroundTasks.Status(tid) == "done")
        break;
    Thread.Sleep(20);
}
Console.WriteLine($"          status={BackgroundTasks.Status(tid)}  count={count}");

// 2) Delayed start, then let it finish
var delayed = BackgroundTasks.SpawnAfter(150, Work);
Console.WriteLine($"after     id={delayed}  status={BackgroundTasks.Status(delayed)}");
for (var i = 0; i < 50; i++)
{
    if (count >= 2 && BackgroundTasks.Status(delayed) == "done")
        break;
    Thread.Sleep(20);
}
Console.WriteLine($"          status={BackgroundTasks.Status(delayed)}  count={count}");

// 3) Cancel before the delay elapses
var doomed = BackgroundTasks.SpawnAfter(500, Work);
Console.WriteLine($"cancel    id={doomed}  cancelled={BackgroundTasks.Cancel(doomed)}");
Thread.Sleep(200);
Console.WriteLine($"          status={BackgroundTasks.Status(doomed)}  count={count} (unchanged)");
